# Hidden Markov Model (HMM) for Named-Entity  Recognition (NER) : Model, inference and learning


# Goals : 
1. A presentation of HMM
2. Formal definitions on inference and learning with HMM
3. An application Named-Entity  Recognition (NER)

# Author: Romain Raveaux (romain.raveaux@univ-tours.fr)

# The lecture
The content of the is notebook is based on the following lectures: 
Supervised Machine Learning for structured input/output: 

*   1\. Introduction to supervised Machine Learning: A probabilistic introduction [PDF](http://romain.raveaux.free.fr/document/courssupervisedmachinelearningRaveaux.pdf)

*   2\. **Connecting local models : The case of chains** [PDF slides](http://romain.raveaux.free.fr/document/Connecting%20local%20models%20the%20case%20of%20chains%20.pdf)

*   3\. Connecting local models : Beyond chains and trees.[PDF slides](http://romain.raveaux.free.fr/document/Structured%20Output%20Learning.pdf)

*   4\. Machine Learning and Graphs : Introduction and problems [PDF slides](http://romain.raveaux.free.fr/document/cours%20IA%20DI5%20graphs%20introV2.pdf)

*   5\. Graph Neural Networks. [PDF slides](http://romain.raveaux.free.fr/document/graph%20neural%20networks%20romain%20raveaux.pdf)

*   6\. Graph Kernels. [PDF slides](http://romain.raveaux.free.fr/document/graph%20kernel%20romain%20raveaux.pdf)


# Named-Entity  Recognition (NER)

Definition taken from Wikipedia (https://en.wikipedia.org/wiki/Named-entity_recognition).

Named-entity recognition (NER) (also known as (named) entity identification, entity chunking, and entity extraction) is a subtask of information extraction that seeks to locate and classify named entities mentioned in unstructured text into pre-defined categories such as person names, organizations, locations, medical codes, time expressions, quantities, monetary values, percentages, etc.

Most research on NER systems has been structured as taking an unannotated block of text, such as this one:

Jim bought 300 shares of Acme Corp. in 2006.

And producing an annotated block of text that highlights the names of entities:

[Jim]Person bought 300 shares of [Acme Corp.]Organization in [2006]Time.

In this example, a person name consisting of one token, a two-token company name and a temporal expression have been detected and classified.

# Problem definition : 
$\mathcal{D}= \{x^{(j)},y^{(j)}\}_{j=1}^N$ with $N$ the number of pair samples.

$x^{(j)} \in \mathcal{S}^{M \times 1}$ is a sequence of $M$ words. A word takes its value in the set $\mathcal{S}$.

$\mathcal{S}$ is the set of all the words. $\mathcal{S}$ can be called a dictionary.

$x^{(j)}_i \in \mathcal{S}$ is a word. $x_i$ for short when there is no need to specify the index of the sequence.

Let us define $D=|\mathcal{S}|$. $D$ is the size of the dictionary. The number of possible words.

$y^{(j)} \in \mathcal{E}^{M \times 1}$ is a sequence of named-entities. A Named-entity takes its value in the set $\mathcal{E}$. 

$\mathcal{E}$ is the set of all the named-entities. 

$y^{(j)}_i \in \mathcal{E}$ is a named-entity. $y_i$ for short when there is no need to specify the index of the sequence.

Let us define $K=|\mathcal{E}|$. $K$ is the size of the named-entity set. The number of named-entities.

### We want to find a function $f: \mathcal{S}^{M \times 1} \to \mathcal{E}^{M \times 1}$.


# A sequence model : HMM
Let us try to model our sequence thanks to a HMM model. 
![Graph Convolution](http://romain.raveaux.free.fr/document/hmmmodel.PNG)

# Inference with HMM : Maximum a posteriori
We take a new set of measurements ($x^{new}$) and use the model to tell us about the world state ($y$).
The HMM is generative model because it models $Pr(x_i|y_i)$ where we want $Pr(y_i|x_i)$.
Inference with a generative model can be achieved by the baye’s rule. We obtain the so called posterior distribution. 
![Graph Convolution](http://romain.raveaux.free.fr/document/MAPHMM.PNG)
Finding the maximum of the posterior distribution is known as Maximum a posteriori (**MAP**).

# Inference with HMM : the optimization problem
Let us define the optimization problem to find the MAP.
$$Pr(y_1,\cdots,y_M|x^{new}_1,\cdots,x^{new}_M)=\frac{Pr(x_1^{new},\cdots,x_M^{new}|y_1,\cdots,y_M).Pr(y_1,\cdots,y_M)}{Pr(x_1^{new},\cdots,x_M^{new})}$$
$$\hat{y_1},\cdots,\hat{y_M}=arg \max_{y_1,\cdots,y_M} \bigg[Pr(y_1,\cdots,y_M|x^{new}_1,\cdots,x^{new}_M) \bigg]$$
$$\hat{y_1},\cdots,\hat{y_M}=arg \max_{y_1,\cdots,y_M} \bigg[ Pr(x_1^{new},\cdots,x_M^{new}|y_1,\cdots,y_M).Pr(y_1,\cdots,y_M) \bigg]$$

$$\hat{y_1},\cdots,\hat{y_M}=arg \max_{y_1,\cdots,y_M} \bigg[ \prod_{i=1}^M Pr(x_i^{new}|y_i) .Pr(y_1)\prod_{i=2}^M Pr(y_i|y_{i-1}) \bigg]$$

$$\hat{y_1},\cdots,\hat{y_M}=arg \min_{y_1,\cdots,y_M} - \log \bigg[ Pr(x_1^{new},\cdots,x_M^{new}|y_1,\cdots,y_M).Pr(y_1,\cdots,y_M) \bigg]$$

$$\hat{y_1},\cdots,\hat{y_M}=arg \min_{y_1,\cdots,y_M} \bigg[ -\sum_{i=1}^M Pr(x_i^{new}|y_i) - Pr(y_1) - \sum_{i=2}^M Pr(y_i|y_{i-1}) \bigg]$$
$$\hat{y_1},\cdots,\hat{y_M}=arg \min_{y_1,\cdots,y_M} \bigg[  \sum_{i=1}^M U_i(y_i) + \sum_{i=2}^M P_i(y_i,y_{i-1}) \bigg]$$
$$U_i(y_i) =-\log [ Pr(x_i^{new}|y_i) ]$$
$$P_i(y_i,y_{i-1}) =-\log [ Pr(y_i|y_{i-1}) ]$$

# Solving the optimization problem
The MAP problem can be solved thanks to the Viterbi Algorithm which complexity is $O(MK^2)$ where $K$ is the number of states for $y_{i,j}$ variables. 

## Viterbi Algorithm
The algorithm' concept :
$S_{1,k}=U_1(y_1=k,x^{new}_1)$

$S_{2,k}=U_2(y_2=k,x^{new}_2)+ \min_l [S_{1,l}+P_2(y_1=l,y_2=k)] $

$S_{i,k}=U_i(y_i=k,x^{new}_i)+ \min_l [S_{i-1,l}+P_i(y_{i-1}=l,y_i=k)] $

$\hat{y}_{M}=argmin_k [S_{M,k}] $

# Learning with HMM
##  Modeling the distributions
So far there is no learning.
1. We just give a sequence (x) and output the labeled sequence (y)
2. Where learning can be introduced ?
3. Where are the parameters ?

Supervised learning
1. Relatively simple. We first isolate the part of the model that we want to learn : $Pr(x_i|y_i;\lambda)$ and $Pr(y_i|y_{i-1};\gamma)$. For example, we might learn the parameters $\lambda$, $\gamma$ from paired examples of $x_i$ and $y_i$. 

2. We model $Pr(x_i|y_i)$ as being categorically distributed, where the parameters depend $y_i$, so that $Pr(x_i|y_i)=Cat_{x_i}[\lambda]$. Categorical distribution is well suited for words because words are by definition categorial values.  $\lambda \in [0, 1]^{K \times D}$. 

3. We model $Pr(y_i|y_{i-1})$ as being categorically distributed, where the parameters depend on the previous sign $y_{i-1}$
so that $Pr(y_i|y_{i-1}=k)=Cat_{y_i}[\gamma]$.  In our case, $\gamma \in [0, 1]^{K \times K}$.

### Recall on the categorical distribution
$y_i \in \mathcal{E}$ is a named-entity. K=$|\mathcal{E}|$ is the number of named-entities.
$Pr(y_i)=Cat_{y_i}[\theta]$
The probabilities of observing the K outcomes are held in a $K \times 1$ parameter vector $\theta = [ \theta_1, \theta_2, \cdots, \theta_K]$, where $\theta_k \in [0, 1]$ and $\sum_{k=1}^K \theta_k =1$

The categorical distribution can be visualized as a normalized histogram with K bins and can be
written as

$Pr(y_i = k) = \theta_k $

For short, we use the notation

$Pr(y_i) = Cat_{y_i}[\theta]$

Alternatively, we can think of the data as taking values $y_i \in \{e^{(1)}, \cdots, e^{(k)} \}$ where $e^{(k)}$ is the $k^{th}$ unit vector; all elements of $e^{(k)}$ are zero except the $k^{th}$, which is one (i.e. $e^{(2)} = [0,1,0,0,0]$ and $e^{(2)}_1=0$). 

Here we can write:

$Pr(y_i = e^{(k)}) = \prod_{j=1}^K \theta_j^{e_j^{(k)}}=\theta_k$
where $e_j^{(k)}$ is the $j^{th}$ element of $e^{(k)}$.

### A non-positional model
We model $Pr(x_i|y_i)$ as we would do it in Natural Language Processing or Computer Vision fields. It means that we are thinking in terms of a time invariant model. $Pr(x_i|y_i)$ does not depends on the time step.


### Limitations of the HMM : 
1. Markov Assumption : The future depends only on the present. Also known as firs order Markov chain
2. It is a generative model. We want to predict $y_i$ and instead we generate $x_i$
3. $Pr(x_i|y_i)$ and $Pr(y_i|y_{i-1})$ holds parameters that are shared through time. All time steps have the same parameters. Corresponding to the assumption of a stationary time series.

## Learning HMM by Maximum Likelihood
Let us denote $W= \{\lambda, \gamma \}$


1 . Let us express the joint probability of the data set 𝒟 . For one input and one output, the HMM model is :
$$Pr(y_1,\cdots,y_M;x_1,\cdots,x_M)=Pr(x_1,\cdots,x_M|y_1,\cdots,y_M).Pr(y_1,\cdots,y_M)$$
$$Pr(y_1,\cdots,y_M;x_1,\cdots,x_M|W)= \bigg[ \prod_{i=1}^M Pr(x_i|y_i,\lambda) \bigg] \bigg[ Pr(y_1)\prod_{i=2}^M Pr(y_i|y_{i-1}, \gamma)\bigg] $$

2 . For the whole data set : Assuming each data sequence was drawn independently from the distribution (Independent and identically distributed (i.i.d) ) !!!

$$Pr(\mathcal{D}|W)=  Pr(y^{(1)},\cdots,y^{(N)};x^{(1)},\cdots,x^{(N)}|W)$$
$$Pr(\mathcal{D}|W)= \prod_{j=1}^N\bigg[Pr(y^{(j)}_1,\cdots,y^{(j)}_M;x^{(j)}_1,\cdots,x^{(j)}_M|W)\bigg]$$
$$Pr(\mathcal{D}|W)= \prod_{j=1}^N\bigg[ \prod_{i=1}^M \bigg( Pr(x^{(j)}_i|y^{(j)}_i,\lambda) \bigg)  Pr(y^{(j)}_1)\prod_{i=2}^M \bigg( Pr(y^{(j)}_i|y^{(j)}_{i-1}, \gamma)\bigg)\bigg] $$
$$\hat{W}=arg \max_W  Pr(\mathcal{D}|W) $$
$$\hat{W}=arg \max_W   \prod_{j=1}^N\bigg[
Pr(y^{(j)}_1,\cdots,y^{(j)}_M;x^{(j)}_1,\cdots,x^{(j)}_M|W)\bigg] $$

3 . Let us fit the probability model by the maximum of likelihood

$$\hat{W}=arg \max_W  \prod_{j=1}^N\bigg[ \prod_{i=1}^M \bigg( Pr(x^{(j)}_i|y^{(j)}_i,\lambda) \bigg)  Pr(y^{(j)}_1)\prod_{i=2}^M \bigg( Pr(y^{(j)}_i|y^{(j)}_{i-1}, \gamma)\bigg)\bigg] $$

$$\hat{\lambda}=arg \max_{\lambda}  \prod_{j=1}^N\bigg[ \prod_{i=1}^M \bigg( Pr(x^{(j)}_i|y^{(j)}_i,\lambda) \bigg)  \bigg]$$


$$\hat{ \gamma}=arg \max_{ \gamma}  \prod_{j=1}^N\bigg[ Pr(y^{(j)}_1)\prod_{i=2}^M \bigg( Pr(y^{(j)}_i|y^{(j)}_{i-1}, \gamma)\bigg)  \bigg]$$

## Maximum Likelihood of a categorical distribution
We estimate the parameters $\lambda=[0,1]^{K \times D}$. $\lambda$ is a matrix. $\lambda_k$ is a vector $\lambda_k=[0,1]^{1 \times D}$ representing the word distribution for a given named-entity $Pr(x_i|y_i=k)=\lambda_k$.

To find the maximum likelihood solution, we maximize the product of the likelihoods for each individual data point with respect to the parameters $\lambda$.

$\hat{\lambda_k}= arg \max_{\lambda_k \in [0, 1]^{D}}  \prod_{j=1}^N\bigg[ \prod_{i=1}^M \bigg( Pr(x^{(j)}_i|y^{(j)}_i=k,\lambda_k) \bigg)  \bigg] \text{  s.t.  }  \sum_{d=1}^D \lambda_{k,d} =1$


We set the derivatives equal to zero, and solve for $\lambda_{k,d}$ to obtain :
$\hat{\lambda_{k,d}}=\frac{N_d}{\sum_{l=1}^{D} N_l}=\frac{Pr(x=d,y=k)}{Pr(y=k)}$

$N_d$ = For a given named-entity (k), the number of time we observe the word $d$.

$\sum_{l=1}^{D} N_l$ = For a given named-entity (k), the sum of the number of time we observe the word $l$.

In other words, $\lambda_{k,d}$ is the proportion of times that we observed the bin $d$.

# Data Set

We need a data set. We will use the CoNLL-2003 databse.
The shared task of CoNLL-2003 concerns language-independent named entity recognition. We will concentrate on four types of named entities: persons, locations, organizations and names of miscellaneous entities that do not belong to the previous three groups.

The CoNLL-2003 shared task data files contain four columns separated by a single space. Each word has been put on a separate line and there is an empty line after each sentence. The first item on each line is a word, the second a part-of-speech (POS) tag, the third a syntactic chunk tag and the fourth the named entity tag. The chunk tags and the named entity tags have the format I-TYPE which means that the word is inside a phrase of type TYPE. Only if two phrases of the same type immediately follow each other, the first word of the second phrase will have tag B-TYPE to show that it starts a new phrase. A word with tag O is not part of a phrase. Note the dataset uses IOB2 tagging scheme, whereas the original dataset uses IOB1.

For more details see https://www.clips.uantwerpen.be/conll2003/ner/ and https://www.aclweb.org/anthology/W03-0419



WORD      POS   CHUNK NE

U.N.      NNP   I-NP  I-ORG

official  NN    I-NP  O

Ekeus     NNP   I-NP  I-PER

heads     VBZ   I-VP  O

for       IN    I-PP  O

Baghdad   NNP   I-NP  I-LOC

.         .     O     O


In [1]:
!wget https://data.deepai.org/conll2003.zip


'wget' n’est pas reconnu en tant que commande interne
ou externe, un programme exécutable ou un fichier de commandes.


# Let us code

## Let us get in touch with data 
### Let us parse the data


In [2]:
# Let us parse the data 
from nltk.corpus.reader import ConllCorpusReader
train = ConllCorpusReader('./dataset', 'eng.train', ['words','pos','ne','chunk'])
allsentences=train.iob_sents()
print("The number of sentences: ",len(allsentences))
print("A sentence: ",allsentences[0])
print("A word and the POS tag and the NER tag : ",allsentences[0][0])
print("POS = Part of Speech")


The number of sentences:  14987
A sentence:  [('EU', 'NNP', 'I-ORG'), ('rejects', 'VBZ', 'O'), ('German', 'JJ', 'I-MISC'), ('call', 'NN', 'O'), ('to', 'TO', 'O'), ('boycott', 'VB', 'O'), ('British', 'JJ', 'I-MISC'), ('lamb', 'NN', 'O'), ('.', '.', 'O')]
A word and the POS tag and the NER tag :  ('EU', 'NNP', 'I-ORG')
POS = Part of Speech


### Let us compute K the number of NER signs and D the number of unique words

In [3]:
setne=set([])
setword=set([])
nbwords=0
for i in range(len(allsentences)):
    onesentence=allsentences[i]
    for j in range(len(onesentence)):
        onetoken=onesentence[j]
        word=onetoken[0]
        pos=onetoken[1]
        ne=onetoken[2]
        #print(word)
        #print(pos)
        #print(ne)
        setword.add(word)
        setne.add(ne)
        nbwords=nbwords+1


K=len(setne)
D=len(setword)
print("K=",K)
print("D=",D)
print("Number of words=",nbwords)

dicne={}
for ne in setne:
    dicne[ne]=len(dicne)

print(dicne)

dicword={}
for word in setword:
    dicword[word]=len(dicword)

K= 8
D= 23623
Number of words= 203621
{'B-LOC': 0, 'B-MISC': 1, 'I-PER': 2, 'I-ORG': 3, 'I-LOC': 4, 'O': 5, 'I-MISC': 6, 'B-ORG': 7}


### Let us compute the probability distribution $Pr(y_i)=Cat_{x_i}[\theta]$.  $\theta \in [0,1]^{1 \times K}$. 

In [4]:
import numpy as np

thetaa=np.zeros((1,K))
for i in range(len(allsentences)):
    onesentence=allsentences[i]
    for j in range(len(onesentence)):
        onetoken=onesentence[j]
        word=onetoken[0]
        pos=onetoken[1]
        ne=onetoken[2]
        indexne=dicne[ne]
        thetaa[0,indexne]=thetaa[0,indexne]+1
        
print(thetaa)
print("We should normalize the thetaa to sum to one but I need it not normalize Pr(x_i|y_i) later")


[[1.10000e+01 3.50000e+01 1.11280e+04 1.00010e+04 8.28600e+03 1.69578e+05
  4.55800e+03 2.40000e+01]]
We should normalize the thetaa to sum to one but I need it not normalize Pr(x_i|y_i) later


### Let us compute the probability distribution $Pr(x_i|y_i)=Cat_{x_i}[\lambda]$.  $\lambda \in [0,1]^{K \times D}$. 


In [5]:
lambdaa=np.zeros((K,D))
for i in range(len(allsentences)):
    onesentence=allsentences[i]
    for j in range(len(onesentence)):
        onetoken=onesentence[j]
        word=onetoken[0]
        pos=onetoken[1]
        ne=onetoken[2]
        indexne=dicne[ne]
        indexword=dicword[word]
        lambdaa[indexne,indexword]=lambdaa[indexne,indexword]+1


for k in range(K):
    lambdaa[k,:]/=thetaa[0,k]
print(lambdaa.sum(axis=1))

[1. 1. 1. 1. 1. 1. 1. 1.]


### Let us compute the probability distribution $Pr(y_i|y_{i-1}=k)=Cat_{y_i}[\gamma]$.  $\gamma \in \mathbb{R}^{K \times K}$.

In [6]:
gammaa=np.zeros((K,K))

for i in range(len(allsentences)):
    onesentence=allsentences[i]
    for j in range(len(onesentence)-1):
        onetoken=onesentence[j]
        ne=onetoken[2]
        indexnejminusone=dicne[ne]
        
        onetoken=onesentence[j+1]
        ne=onetoken[2]
        indexnej=dicne[ne]
        gammaa[indexnejminusone,indexnej]=gammaa[indexnejminusone,indexnej]+1

for k in range(0,K):
    gammaa[k,:]=gammaa[k,:]/gammaa[k,:].sum()
print(gammaa.sum(axis=1))

[1. 1. 1. 1. 1. 1. 1. 1.]


# Let us code the Viterbi Algorithm

The algorithm' concept :
$S_{1,k}=U_1(y_1=k,x^{new}_1)$

$S_{2,k}=U_2(y_2=k,x^{new}_2)+ \min_l [S_{1,l}+P_2(y_1=l,y_2=k)] $

$S_{i,k}=U_i(y_i=k,x^{new}_i)+ \min_l [S_{i-1,l}+P_i(y_{i-1}=l,y_i=k)] $

$\hat{y}_{M}=argmin_k [S_{M,k}] $


In [7]:

def Ui(yi,x):
    epsilon = 0.0000000001
    pr=-np.log(lambdaa[yi,x]+epsilon)
    return pr

def Pi(yi,yiminusone):
    epsilon = 0.0000000001
    return -np.log(gammaa[yiminusone,yi]+epsilon)

def FindMin(S,yi,i):
    minindex=-1
    minval=999999
    for l in range(0,K):
        vals=S[i-1,l]
        valp=Pi(yi,l)
        val=vals+valp
        if val < minval:
            minval=val
            minindex=l
    return minval,minindex

def Viterbi(xnew):
    sentencelength=len(xnew)

    S=np.zeros((sentencelength,K))
    Sindice=np.zeros((sentencelength,K))
    i=0
    x=xnew[i]
    for yi in range(0,K):
        S[i,yi]=Ui(yi,x)
    
    for i in range(1,sentencelength):
        x=xnew[i]
        for yi in range(0,K):
            minvall,minl=FindMin(S,yi,i)
            Sindice[i,yi]=minl
            S[i,yi]=Ui(yi,x)+minvall

    ypred=[]
    ycur=S[sentencelength-1,:].argmin()
    ypred.append(ycur)
    i=sentencelength-1
    while i >= 1:
        ycur=Sindice[i,int(ycur)]
        ypred.append(int(ycur))
        i=i-1
    ypred.reverse()
    return ypred
    

## Let us try the algorithm

In [8]:
onesentence=allsentences[0]
print(onesentence)
print("Let us get only the words")
words=[]
for mytuple in onesentence:
    words.append(mytuple[0])
print(words)
print("Let us transform the words to integers")
wordsinteger=[]
for word in words:
    wordsinteger.append(dicword[word])
print(wordsinteger)

print("Let us call Viterbi")
ypred=Viterbi(wordsinteger)
print(ypred)


print("Let us get only the named-entities")
ywords=[]
for mytuple in onesentence:
    ywords.append(mytuple[2])
print(ywords)
print("Let us transform thenamed-entities to integers")
y=[]
for yword in ywords:
    y.append(dicne[yword])
print(y)

print("Let us compare y and ypred by a ratio of similarity")
score=np.sum(np.array(y)==np.array(ypred))/len(y)
print(score)


[('EU', 'NNP', 'I-ORG'), ('rejects', 'VBZ', 'O'), ('German', 'JJ', 'I-MISC'), ('call', 'NN', 'O'), ('to', 'TO', 'O'), ('boycott', 'VB', 'O'), ('British', 'JJ', 'I-MISC'), ('lamb', 'NN', 'O'), ('.', '.', 'O')]
Let us get only the words
['EU', 'rejects', 'German', 'call', 'to', 'boycott', 'British', 'lamb', '.']
Let us transform the words to integers
[19162, 2832, 22573, 12362, 17509, 21542, 6855, 3829, 15450]
Let us call Viterbi
[3, 5, 6, 5, 5, 5, 6, 5, 5]
Let us get only the named-entities
['I-ORG', 'O', 'I-MISC', 'O', 'O', 'O', 'I-MISC', 'O', 'O']
Let us transform thenamed-entities to integers
[3, 5, 6, 5, 5, 5, 6, 5, 5]
Let us compare y and ypred by a ratio of similarity
1.0


# Let us write some helpers 

In [9]:
def sentencetoword(onesentence):
    words=[]
    for mytuple in onesentence:
        words.append(mytuple[0])
    return words

def wordstointegers(words):
    wordsinteger=[]
    for word in words:
        wordsinteger.append(dicword[word])
    return wordsinteger

def sentencetoNE(onesentence):
    ywords=[]
    for mytuple in onesentence:
        ywords.append(mytuple[2])
    return ywords

def NEtointegers(ywords):
    y=[]
    for yword in ywords:
        y.append(dicne[yword])
    return y

def similaritymeasure(y,ypred):
    score=np.sum(np.array(y)==np.array(ypred))/len(y)
    return score
    


# Let us try our method on all the database

In [10]:
allscore=0
count=0
for i in range(len(allsentences)):
    #print(i)
    onesentence=allsentences[i]
    
    if(len(onesentence)>1):    
        words=sentencetoword(onesentence)
        wordsinteger=wordstointegers(words)

        ywords=sentencetoNE(onesentence)
        y=NEtointegers(ywords)


        ypred=Viterbi(wordsinteger)
        score=similaritymeasure(y,ypred)
        allscore+=score
        count=count+1

allscore/=count
print(allscore)

0.9811974191082682


### We get 98% of good named-entity recognition on the training data set. !!!! Well done

# Conclusion

1. We have implemented a Hidden Markov Model for named-entity recognition.
2. The learning phase is achieved by modeling probability distributions as categorical distributions and parameters of these distructions are learned by Maximum Likelihood.
3. Inference is achieved by Maximum a posteriori thanks to the Viterbi algorithm

